### PR TITLE
Configuration Handling Refactor

### PR DESCRIPTION
**Pull request by elqurio**
## Description

Due to the current structure and the difference between the config that is provided, and the config that is actually consumed (and how it's consumed), I had to create transformers for the json data. Their purpose is to transform the provided json config and turn it into a structure that pydantic models can consume without issues.

I think our goal should be to keep the structure provided as close as possible to the structure that is consumed, just to make things more straightforward. I'd like to believe that we can completely remove transformers, and consume data in the same structure it is provided.

This PR should bring more clarity into what is really happening during the configuration loading, making it much easier to change and further improve.

Also, typing support is never a bad thing, plus we can export the json schema of the models which is an additional benefit.
## Type of Change

- [ ] Bug Fix
- [ ] New Feature
- [ ] Breaking Change
- [x] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [ ] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [ ] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: lib/python/flame/config.py [hunk 1]
```diff
@@ -1,433 +1,258 @@
-# Copyright 2022 Cisco Systems, Inc. and its affiliates
```
Comment (myungjin) [inline lib/python/flame/config.py:1]: Please keep the copyright.

File: lib/python/flame/config.py [hunk 2]
```diff
@@ -1,433 +1,258 @@
...
+    dataset: str
+    max_run_time: int
+    base_model: BaseModel
+    groups: t.Optional[Groups]
+    dependencies: list[str]
+    func_tag_map: t.Optional[dict]
+
+
+def read_config() -> dict:
+    with open("flame/examples/mnist/trainer/config.json") as f:
```
Comment (myungjin) [inline lib/python/flame/config.py:152]: Why is a filename hard-coded? Is this for testing purpose? If this is intended, please leave a comment. If not, please add `TODO`.

**Review by myungjin — COMMENTED** [review]
left two comments. please consider to address them.

File: lib/python/flame/config.py [hunk 3]
```diff
@@ -1,433 +1,258 @@
...
-# See the License for the specific language governing permissions and
-# limitations under the License.
-#
-# SPDX-License-Identifier: Apache-2.0
-"""Config parser."""
-
-import json
-import sys
 from enum import Enum
+from pydantic import Field
```
Comment (myungjin) [inline lib/python/flame/config.py:2]: If `pydantic` is not one of default python packages, please update dependencies in `setup.py` (i.e., `install_requires`).

File: lib/python/flame/config.py (hunk 2 above)
Comment (elqurio) [reply lib/python/flame/config.py]: Fixed this, it accepts an argument filename which is passed during initialisation.

File: lib/python/flame/config.py (hunk 3 above)
Comment (elqurio) [reply lib/python/flame/config.py:19]: Added, nice catch!

**Review by myungjin — APPROVED** [review]
lgtm

**Pull request merged**